TESTed (#124)
dodona-edu/dodona-edu.github.io: en/guides/developers/tested-configure-new-programming-language/README.md

---
title: "[nl] Configureren nieuwe programmeertaal TESTed"
description: "[nl] Configureren nieuwe programmeertaal TESTed"
---

::: warning Sorry
For now, this page is only available in Dutch. Sorry!
:::

# Configuratie van een programmeertaal
In deze handleiding wordt in detail uitgelegd hoe een nieuwe programmeertaal aan TESTed kan toegevoegd worden.
We doen dat door te beschrijven hoe de programmeertaal C toegevoegd is.
Enkele nuttige links die hierbij kunnen helpen zijn:

- Bestaande configuraties: <https://github.com/dodona-edu/universal-judge/tree/master/tested/languages>
- Testoefeningen: <https://github.com/dodona-edu/universal-judge/tree/master/exercise>

In deze referentie worden regelmatig codebestanden getoond, verspreid over meerdere codefragmenten
(bijvoorbeeld eerst `n` regels, daarna wat tekst en dan pas de rest van de regels).

## TESTed lokaal uitvoeren
Tijdens het configureren van een programmeertaal is het nuttig om TESTed lokaal uit te voeren,
zonder daarvoor het volledige Dodona-platform te moeten uitvoeren.
Buiten de _dependencies_ voor de bestaande programmeertalen is TESTed een Python-package,

> dat op de normale manier uitgevoerd kan worden.

Author: I think we want assume to assume that, because when adding a programming language TESTed, we expects some knowledge about Python.
This information is also found in the README of the TESTed repo.

  Door optimalisaties is het momenteel aan te raden om dezelfde Python-versie te gebruiken voor als voor de
  Python-oefeningen.

- **Java**:
  vereist Java 11, maar heeft verder geen dependencies.
  De commando's `javac` en `java` moeten beschikbaar zijn in het `PATH`.

- **JavaScript**:
  vereist NodeJS v10, maar heeft verder geen dependencies.
  Het commando `node` moet beschikbaar zijn in het `PATH`.

- **Kotlin**:
  vereist Kotlin 1.4.10, maar heeft verder geen dependencies.
  De commando's `kotlinc` en `kotlin` moeten beschikbaar zijn in het `PATH`.

- **Haskell**:
  Voor Haskell is GHC 8.6 (`ghc`) of later nodig.
  Daarnaast is `aeson` nodig.
  Beiden moeten globaal beschikbaar zijn in het `PATH`.

Merk op dat de dependencies voor de programmeertalen optioneel zijn.
Om bijvoorbeeld enkel Python-oplossingen te beoordelen zijn geen andere dependencies nodig.

Voor de programmeertaal C gaan we gebruikmaken van GCC, waarbij versie 8.1 of later nodig is.

TESTed werkt op elk besturingssysteem dat ondersteund wordt door Python.
Sommige dependencies, zoals GCC, vragen wel meer moeite om te installeren op Windows.

::: tip Windows tip

> Gebruikers op Windows kunnen MingW of MSYS2 proberen.

Author: That should only be recommended for Windows 10 Users

"int16": "short",
"uint16": "unsigned short",
"int32": "int",
"uint32": "unsigned int",
"int64": "long",
"uint64": "unsigned long",
"bigint": "long long",
"single_precision": "float",
"double_precision": "double",
"any": "void*",
"inner": {
},
```

Ondersteunde waarden:
- **string**: Naam van het TESTed datatype in de programmeertaal.
- **boolean**: Alleen voor collectie datatypes waarbij de notatie van het collectietype uitsluitend met haakjes gebeurt.
  - `true`: Het datatype van de gegevens in de collectie bevindt zich tussen de haakjes.
    Hierbij moeten de haakjes van het collectie type expliciet worden opgeven.
    Bijvoorbeeld Haskell lijst: `[Int]`
  - `false`: De haakjes bevinden zich achter het datatype van de gegevens in de collectie.
    Hierbij moeten de haakjes van het collectie type expliciet worden opgeven.
    Bijvoorbeeld Java array: `int[]`

`inner` is het object waarbij elk datatype, die een andere naam moet krijgen in een collectie, moet opgegeven worden.
Dit geldt niet voor datatypes waarbij de haakjes achter het datatype geplaatst worden.
Bijvoorbeeld Java: `Integer` in plaats van `int` voor `int32`.

#### Natuurlijke taal constructies

> De benaming in de natuurlijke taal voor de collectie datatypes en tekst, is afhankelijk van de programmeertaal.

Reviewer: ```suggestion
De benaming in de natuurlijke taal voor de collectiedatatypes en tekst, is afhankelijk van de programmeertaal.
```

What is the "tekst" part here? As far as I can see, we only specify data type names in the section of the guide.

Hiervoor wordt gevraagd om deze benaming op te geven in zowel het Engels als het Nederlands.

> De vereiste velden zijn alle collectie datatypes die ondersteund worden in TESTed door de programmeertaal,

Reviewer: ```suggestion
De vereiste velden zijn alle collectiedatatypes die ondersteund worden in TESTed door de programmeertaal,
```

Would it be an idea to have a default fallback, e.g. the type itself? E.g. if nothing is specified for a `sequence`, we fall back to `sequence`.

Author: Good idea, for `type_name` this is default behavior, but not for `natural_type_name`.

alsook de benaming voor tekst.

Bijvoorbeeld: in JavaScript spreekt men over *array* voor een `sequence`,
terwijl in Python men spreekt over een *lijst*.

```json
"natural": {
  "en": {
    "text": "string"
  },
  "nl": {
    "text": "string"
  }
}
```

### Configuratieklasse

> De configuratieklasse is de schakel tussen de generieke aspecten van en het programmeertaalafhankelijke gedrag.

Reviewer: ```suggestion
De configuratieklasse is de schakel tussen de generieke aspecten van [?] en het programmeertaalafhankelijke gedrag.
```

> Omdat in Python geschreven is, moet deze klasse ook in Python geïmplementeerd worden.

Reviewer: ```suggestion
Omdat [?] in Python geschreven is, moet deze klasse ook in Python geïmplementeerd worden.
```

class C(Language):
```

In de rest van deze paragraaf overlopen we de verschillende methodes die geïmplementeerd moeten worden in deze klasse.
In de superklasse, `Language`, zijn de abstracte methodes voorzien van uitgebreide documentatie.

#### Compileren van de code
Een eerste en belangrijke methode is de callback voor de compilatiestap:
```python
def compilation(self, config: Config, files: List[str]) -> CallbackResult:
    main_file = files[-1]
    exec_file = Path(main_file).stem
    result = executable_name(exec_file)
    return (["gcc", "-std=c11", "-Wall", "evaluation_result.c", "values.c",
             main_file, "-o", result], [result])
```

De eerste parameter van deze methode is een klasse met enkele configuratie-opties, zoals de tijdslimiet, geheugenlimiet
en de programmeertaalspecifieke opties.
Dit zou bijvoorbeeld gebruikt kunnen worden om de versie van C mee te geven (zoals C11 of C99).
Dit wordt momenteel niet gedaan in C, want TESTed vereist C11, maar de mogelijkheid bestaat. 
Wel moeten we opmerken dat de tijdslimiet zelden nuttig zal zijn, daar TESTed de uitvoeringstijd bijhoudt.
In de configuratieklasse is het dus niet nodig om daar rekening mee te houden.

> De andere parameter van deze methode is een lijst van bestanden waarvan vermoedt dat ze nuttig kunnen zijn voor de

Reviewer: ```suggestion
De andere parameter van deze methode is een lijst van bestanden waarvan [?] vermoedt dat ze nuttig kunnen zijn voor de
```

Het compilatiecommando neemt de vorm aan van een lijst van de elementen waaruit het commando bestaat.
Bij het uitvoeren van dit commando zal deze lijst aan de Python-module `subprocess` gegeven worden.

Na het uitvoeren van het compilatiecommando moet TESTed weten welke bestanden relevant zijn om mee te nemen naar een
volgende stap in de beoordeling.
Daarom moet een lijst van resulterende bestanden teruggegeven worden.
Enkel bestanden in deze lijst zullen bijvoorbeeld beschikbaar zijn bij het uitvoeren van de runs.
Een lijst van bestanden teruggeven is mogelijk indien op voorhand geweten is in welke bestanden de compilatie
resulteert.
Dit is bijvoorbeeld hier het geval (in C resulteert de compilatie in één uitvoerbaar bestand), of ook bij Python,
waar de compilatie voor elk `.py`-bestand resulteert in een `.pyc`-bestand.
Ook hier moet de conventie gerespecteerd worden dat het bestand met de `main`-functie als laatste komt.

Het is echter niet altijd mogelijk om op voorhand te weten in welke bestanden de code zal resulteren.
Zo resulteert compilatie van één `.java`-bestand mogelijk in meerdere `.class`-bestanden,

> afhankelijk vande inhoud van de bestanden.

Reviewer: ```suggestion
afhankelijk van de inhoud van de bestanden.
```

De filterfunctie krijgt als argument de naam van een bestand en moet `True` of `False` teruggeven als het bestand
respectievelijk wel of niet moet meegenomen worden naar een volgende stap.

Een voorbeeld van de in- en uitvoer van de compilatiemethode:
```python
>>> compilation(['submission.c', 'evaluation_result.c', 'context_0_0.c', 'selector.c'])
(
    ['gcc', '-std=c11', '-Wall', 'evaluation_result.c', 'values.c', 'selector.c',
     '-o', 'selector.exe'], ['selector.exe']
)
```

Als een leeg compilatiecommando wordt teruggegeven, dan wordt er geen compilatie gedaan.
Dit is ook de standaardimplementatie van deze methode.
Voor programmeertalen waar geen compilatie nodig is, moet deze methode niet geïmplementeerd worden.

#### Uitvoeren van de testcode
Na het compileren moeten we een methode implementeren om de gecompileerde code uit te voeren:
```python
def execution(self, config: Config,
              cwd: Path, file: str, arguments: List[str]) -> Command:
    local_file = cwd / executable_name(Path(file).stem)
    return [str(local_file.absolute()), *arguments]
```

Deze functie heeft vier parameters:

- `config`:
  Dezelfde configuratie-opties als bij de compilatiemethode.

>   Bij Java wordt dit bijvoorbeeld gebruikt om de geheugenlimiet van de juist in te stellen.

Reviewer: ```suggestion
  Bij Java wordt dit bijvoorbeeld gebruikt om de geheugenlimiet van de [?] juist in te stellen.
```

- `cwd`:
  de map waarin het uitvoeren plaatsvindt
- `file`:
  het uitvoerbaar bestand dat moet uitgevoerd worden
- `arguments`:
  argumenten die aan het proces moeten meegegeven worden

Als resultaat moet het commando teruggegeven worden, dat ook aan `subprocess` doorgegeven wordt.

In het geval van C is dit commando eenvoudig:
we geven het absolute pad naar het uitvoerbare bestand mee en geven ook de argumenten mee.
Het absolute pad is nodig omdat de executable die we willen uitvoeren
(en gemaakt hebben in de compilatiestap) niet in het `PATH` zit.

Een voorbeeld van deze functie in werking is:
```python
>>> execution('/test/path', 'executable.exe', ['arg1', 'arg2'])
['/test/path/executable.exe', 'arg1', 'arg2']
```

De basisimplementatie van de configuratie is nu klaar.
Voor de meeste programmeertalen kan nu overgegaan worden naar de sjablonen,
maar in C moeten we nog een extra methode implementeren.

#### Aanpassen van de ingediende oplossing

> De testcode die door gegenereerd wordt, kan meerdere `main`-functies bevatten:

Reviewer: ```suggestion
De testcode die door [?] gegenereerd wordt, kan meerdere `main`-functies bevatten:
```

- `run.c`: het sjabloon voor run (**verplicht**) 
- `selector.c`: het sjabloon voor de selector voor batchcompilatie (**verplicht** bij batchcompilatie) 
- `declaration.mako`: vertaalt de declaratie van een variabele naar code
- `function.mako`: vertaalt een functieoproep naar code
- `statement.mako`: vertaalt een statement of een expressie naar code (**verplicht**)
- `value.mako`: vertaalt een letterlijke waarde naar code
- `value_arguments.mako`:
  hulpsjabloon voor `value.mako` (opsomming van recursieve gegevenstypes, zoals lijsten, maar ook van functieargumenten)
- `value_basic.mako`: hulpsjabloon voor `value.mako` (vertaalt de basistypes)

Al deze sjablonen komen in de map `tested/languages/c/templates`.

Het is vrij om de bestandsextensie van de sjablonen te kiezen, zolang het een extensie is uit de configuratie.
Standaard zijn de toegelaten extensies `.mako` en een programmeertaalafhankelijke extensie, hier `.c`.
Een conventie die gebruikt wordt binnen TESTed, is de volgende:
- Sjablonen eindigen op de programmeertaalafhankelijke extensie (`.c`)
  indien het sjabloon resulteert in een op zichzelf staand bestand.

>   Voorbeelden zijn het contextsjabloon en de selector.

Reviewer: Should this be the "run" template?

Author: Good point

Dit wordt niet afgedwongen door TESTed; alle sjablonen hadden de extensie `.c` of `.mako` kunnen krijgen,
of een mengeling.
Dit geldt voornamelijk voor de verplichte sjablonen.
De andere sjablonen (die als hulpmiddel gebruikt worden door de verplichte sjablonen) kunnen eender welke
extensie krijgen, want bij het gebruiken van een sjabloon in Mako moet de bestandsextensie opgegeven worden.

#### Het runsjabloon
Dit is veruit het grootste en het meest ingewikkelde sjabloon.
Het is verantwoordelijk voor het genereren van de testcode voor één run.

We importeren de values-module (hierover later meer) en de ingediende oplossing.
De variabele `submission_name` zal de naam van het oplossingsbestand bevatten.
Een overzicht van alle beschikbare variabelen in het runsjabloon is te vinden in de klasse
`_ExecutionArguments` uit de module `tested.languages.generator`.

We importeren ook alle programmeertaalspecifieke evaluatoren die we nodig zullen hebben.
De variabele `evaluator_names` bevat een verzameling van deze namen.

```c
#include <stdio.h>

#include "values.h"
#include "${submission_name}.c"

% for name in evaluator_names:
    #include "${name}.c"
% endfor
```

##### Witruimte in Mako

> Nuttig om weten is dat een extensie heeft toegevoegd aan Mako,

Reviewer: ```suggestion
Nuttig om weten is dat [?] een extensie heeft toegevoegd aan Mako,
```

```python
LANGUAGES = {
  'c': C,
  'haskell': Haskell,
  'java': Java,
  'javascript': JavaScript,
  'kotlin': Kotlin,
  'python': Python,
  'runhaskell': RunHaskell,
}
```

Om de programmeertaal manueel te testen is volgend stappenplan aanbevolen:
1. Implementeer oplossingen voor een of meerdere oefeningen uit de map `exercises` in de nieuwe programmeertaal.
2. Wijzig `tested/manual.py` zodat dit bestand de oefening gebruikt waarvoor een oplossing bestaat
   (en stel ook de juiste programmeertaal in).
3. Voer uit, zoals we in het begin van het hoofdstuk besproken hebben:
   ```bash
   > python -m tested.manual
   ```

TESTed heeft ook een testsuite met verschillende oefeningen en scenario's.
Om de nieuwe programmeertaal hieraan toe te voegen, moeten de juiste oplossingen geïmplementeerd worden.
Hiervoor wordt best gekeken naar `tests/test_functionality.py`.
In dat bestand staan de verschillende testen.
Bij elke test staat welke oplossing gebruikt wordt; indien het niet duidelijk zou zijn wat de oplossing voor een
bepaalde test moet doen, kunnen de bestaande oplossingen in de bestaande programmeertalen een grote hulp zijn.

> ## Hoe lang duurt het configureren van een programmeertaal?

Reviewer: Is this section relevant for the guide on the website?

configureren van een programmeertaal flexibel te maken, maar dat neemt niet weg dat het configureren van een
programmeertaal grotendeels opgesteld is op basis van de noden van de bestaande programmeertalen.
Hoe meer de programmeertaal afwijkt van de bestaande programmeertalen, des te meer creativiteit en tijd nodig zal zijn
om de programmeertaal in te configureren of om meer hooks en callbacks te voorzien in de configuratieklasse.

Dit alles neemt niet weg dat we optimistisch zijn dat het configureren van een programmeertaal eerder snel kan verlopen.
Gegeven een goede kennis over de programmeertaal (en dat de programmeertaal niet te exotisch is),
schatten we in dat het configureren van een programmeertaal ongeveer een dag in beslag neemt.
Hiermee bedoelen we deze referentie lezen en de stappen die erin beschreven staan uitvoeren, zodat er een werkende
configuratie is.
Om de programmeertaal tot in de puntjes uit te werken voorzien we een kleine twee dagen.
Hiermee bedoelen we bijvoorbeeld het toevoegen van de programmeertaal aan de testen,
waarvoor er verschillende oefeningen opgelost moeten worden.

> ## Stabiliteit van TESTed

Reviewer: Same here: is this relevant for the dodona docs?
